sql-plugin/src/main/scala/com/nvidia/spark/rapids/CostBasedOptimizer.scala

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.expressions.{Alias, AttributeReference, Expression}
+import org.apache.spark.sql.execution.{ProjectExec, SortExec, SparkPlan}
+import org.apache.spark.sql.types.DataTypes
+
+object CostBasedOptimizer extends Logging {
+
+  // the intention is to make the cost model pluggable since we are probably going to need to
+  // experiment a fair bit with this part
+  private val costModel = new DefaultCostModel
+
+  /**
+   * Walk the plan and determine CPU and GPU costs for each operator and then make decisions
+   * about whether operators should run on CPU or GPU.
+   */
+  def optimize(plan: SparkPlanMeta[_]): (Double, Double) = {
+
+    // get the CPU and GPU cost of the child plan(s)
+    val (childCpuCosts, childGpuCosts) = plan.childPlans.map(optimize).unzip
+
+    // get the CPU and GPU cost of this operator
+    val (operatorCpuCost, operatorGpuCost) = costModel.applyCost(plan)
+
+    // calculate total (this operator + children)
+    val totalCpuCost = operatorCpuCost + childCpuCosts.sum
+    var totalGpuCost = operatorGpuCost + childGpuCosts.sum
+
+    // determine how many transitions between CPU and GPU are taking place
+    val numTransitions = plan.childPlans
+      .count(_.canThisBeReplaced != plan.canThisBeReplaced)
+
+    if (numTransitions > 0) {
+      if (plan.canThisBeReplaced) {
+        // transition from CPU to GPU
+        val transitionCost = numTransitions * costModel.transitionToGpuCost(plan)
```

Comment:
I think we need to be clearer on what `transitionToGpuCost` is actually doing.  Right now it is mostly a place holder, but in the future it could look at the schema to determine the cost of transitioning.  We might also want to guess at the size of the data that will transfer. To me it feels like we would want to do something like.

```
plan.childPlans.filter(_.canThisBeReplaced != plan.canThisBeReplaces).map(costModel.transitionToGpuCost(_)).sum
```

So that the transition cost is looking at the output of the plan, so it knows what to transition instead of some vague thing about what the input might be.  Or not pass in the plan at all if we just want it to be a single value and we can change things when the models get better.